mempool/src/pool.rs

```diff
@@ -0,0 +1,903 @@
+    //But we don't yet have a way of summing of the inputs
+    //TODO handle the error bettter
+    fn get_fee(&self) -> Amount {
+        self.tx
+            .get_outputs()
+            .iter()
+            .map(|output| output.get_value())
+            .sum::<Option<_>>()
+            .expect("fee overflow")
+    }
+}
+
+impl PartialOrd for TxMempoolEntry {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(compare_fees(self, other))
+    }
+}
+
+impl Ord for TxMempoolEntry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        compare_fees(self, other)
+    }
+}
+
+pub(crate) struct MempoolImpl<C: ChainStateView> {
+    store: MempoolStore,
+    chain_state: C,
+}
+
+struct MempoolStore {
+    // TODO some cloning can be avoided by having spender_txs be a map between references
+    txs_by_id: HashMap<H256, Rc<TxMempoolEntry>>,
+    txs_by_fee: BTreeSet<Rc<TxMempoolEntry>>,
+    spender_txs: HashMap<OutPoint, Rc<TxMempoolEntry>>,
+}
+
+impl MempoolStore {
+    fn new() -> Self {
+        Self {
+            txs_by_fee: BTreeSet::new(),
+            txs_by_id: HashMap::new(),
+            spender_txs: HashMap::new(),
+        }
+    }
+
+    // Checks whether the outpoint belongs to one of the txs in the mempool
+    fn contains_outpoint(&self, outpoint: &OutPoint) -> bool {
+        matches!(self.txs_by_id.get(&outpoint.get_tx_id().get()),
+            Some(entry) if entry.tx.get_inputs().len() > outpoint.get_output_index() as usize)
+    }
+
+    fn create_entry(&self, tx: Transaction) -> TxMempoolEntry {
+        let parents = tx
+            .get_inputs()
+            .iter()
+            .filter_map(|input| self.txs_by_id.get(&input.get_outpoint().get_tx_id().get()))
+            .cloned()
+            .collect::<BTreeSet<_>>();
+
+        // We implement what is known in BIP125 as "inherited signalling":
+        // A Transaction is replaceable if either (1) it signals replaceability explicitly (via its flags), OR (2) one of its in-mempool ancestors explicit signals replaceability
+
+        // At the level of TxMempoolEntry, signalling replaceability is always explicit, via the
+        // `is_replaceable` field. By construction, to determine the value of this field it is
+        // enoufh to scan direct parents, not all ancestors.
+        let is_replaceable =
+            tx.is_replaceable() || parents.iter().any(|parent| parent.is_replaceable);
+        TxMempoolEntry::new(tx, parents, is_replaceable)
+    }
+
+    fn add_transaction(&mut self, tx: Transaction) -> Result<(), MempoolError> {
+        let id = tx.get_id().get();
+        let entry = self.create_entry(tx);
+        let entry_rc = Rc::new(entry);
+        self.txs_by_id.insert(id, entry_rc.clone());
+        self.txs_by_fee.insert(entry_rc.clone());
+
+        for outpoint in entry_rc.tx.get_inputs().iter().map(|input| input.get_outpoint()) {
+            self.spender_txs.insert(outpoint.clone(), entry_rc.clone());
+        }
```

Comment:
There is a clippy lint for this but it's allowed by default. Should be promote it to a warning?
https://rust-lang.github.io/rust-clippy/master/#clone_on_ref_ptr

Reply:
I love it. I don't believe in warnings though, since people tend to ignore them. Added `#![deny(clippy::clone_on_ref_ptr)]
` at the mempool crate level.